### ISSUE TITLE
Scoped and in_place waitable tasklists

### ISSUE THREAD

**Issue by HeroicKatora** [rayon-rs/rayon#920]
Separated from: https://github.com/image-rs/jpeg-decoder/issues/227

The problem is as followed, as part of decoding work we spawn a bunch of tasks. These tasks must communicate their results back when collected but crucially, tasks do _not_ complete in a lexical manner but as separate groups where sometimes their results must be utilized for further work—with this decision depending on data that is only available _after_ such tasks are being spawned and thus this work happens as a separate task.

This implies it is not feasible to wrap the entire thing in a scope to perform that waiting. The only guaranteed synchronization point is after the _lexcial_ end of the scope. The current situation is that this leads to deadlock: The work coordination happens in the main thread but it can never pick-up any work itself. None of `scope`, `scope_fifo`, `spawn` can be utilized for this as they either (a) do not wait on the tasks that had been spawn non-lexically or (b) can't 'block' to return a value to the scope.

What I'd like to have is an API similar to `Scope`, except I can pre-emptively _consume_ the scope to perform an in-place wait Avoiding a deadlock because the current thread can always be used as a fallback to perform the tasks.
